src/django_otp/models.py

```diff
@@ -186,6 +190,49 @@ def verify_token(self, token):
         return False
 
 
+class SideChannelDevice(Device):
+    """
+    Abstract base model for a side-channel :term:`device` attached to a user.
+    This model implements token generation, verification and expiration, so
+    the concrete devices only have to implement delivery.
+    """
+    token = models.CharField(max_length=16, blank=True, null=True)
+    valid_until = models.DateTimeField(
+        default=timezone.now,
+        help_text="The timestamp of the moment of expiry of the saved token."
+    )
+
+    class Meta:
+        abstract = True
+
+    def generate_token(self, length=6, valid_secs=300):
+        """
+        Generates a token of the specified length, then sets it on the model
+        and sets the expiration of the token on the model.
+
+        The concrete device should save the object after calling this method.
```

Comment:
Similar to ModelForm, ThrottlingMixin automatically commits in situations like this, unless given `commit=False`. Might be more consistent.

Reply:
Good idea, it's better to be consistent.